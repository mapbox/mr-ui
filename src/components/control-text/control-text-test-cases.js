import React from 'react';
import ControlText from './control-text';
import safeSpy from '../../test-utils/safe-spy';

const testCases = {};

testCases.basic = {
  description: 'basic',
  component: ControlText,
  props: {
    id: 'testinput',
    label: 'Basic',
    onChange: safeSpy()
  }
};

testCases.disabled = {
  description: 'disabled',
  component: ControlText,
  props: {
    id: 'testinput',
    label: 'Disabled',
    onChange: safeSpy(),
    disabled: true
  }
};

testCases.validationError = {
  description: 'default validation error',
  component: ControlText,
  props: {
    id: 'testinput',
    label: 'Validation error',
    validationError: 'oh no!',
    onChange: safeSpy()
  }
};

testCases.errorInline = {
  description: 'error inline option',
  component: ControlText,
  props: {
    id: 'testinput',
    label: 'Error inline',
    errorStyle: 'inline',
    validationError: 'oh no!',
    onChange: safeSpy()
  }
};

testCases.allOptions = {
  description: 'all options',
  component: ControlText,
  props: {
    id: 'testinput',
    label: 'All options',
    type: 'email',
    validationError: 'oh no!',
    errorStyle: 'inline',
    popoverProps: {
      coloring: 'dark',
      padding: 'small'
    },
    placeholder: 'Enter an email address',
    value: 'foo@bar.baz',
    optional: true,
    autoFocus: true,
    noAuto: true,
    aside: <span>Aside text</span>,
    themeControlInput: 'txt-bold input',
    themeControlWrapper: 'bg-gray-faint',
    themeLabel: 'txt-s txt-bold',
    onChange: safeSpy()
  }
};

export { testCases };
