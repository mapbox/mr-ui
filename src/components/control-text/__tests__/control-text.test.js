import React from 'react';
import { shallow, mount } from 'enzyme';
import { testCases } from './control-text-test-cases';
import getWindow from '../../utils/get-window';

jest.mock('../../utils/get-window', () => {
  return jest.fn();
});

// Mocking react-popover as Portals in React 16 (which `react-displace` uses if
// available) is not currently supported in enzyme. see
// https://github.com/airbnb/enzyme/issues/1150

/* eslint-disable */
jest.mock('../../popover', () => () => {
  return <div data-test="popover-component" />;
});
/* eslint-enable */

describe('ControlText', () => {
  let testCase;
  let wrapper;

  describe(testCases.basic.description, () => {
    beforeEach(() => {
      testCase = testCases.basic;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('onChange gets called with input id and value', () => {
      const mockEvent = {
        target: {
          value: 'foo'
        }
      };
      wrapper.find('input').props().onChange(mockEvent);
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      expect(testCase.props.onChange).toHaveBeenCalledTimes(1);
      expect(testCase.props.onChange).toHaveBeenCalledWith('foo', 'testinput');
    });
  });

  describe(testCases.disabled.description, () => {
    beforeEach(() => {
      testCase = testCases.disabled;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('disabled is true', () => {
      const props = wrapper.find('input').last().props();
      expect(props).toHaveProperty('disabled', true);
    });
  });

  describe(testCases.validationError.description, () => {
    beforeEach(() => {
      testCase = testCases.validationError;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe(testCases.errorInline.description, () => {
    beforeEach(() => {
      testCase = testCases.errorInline;
      wrapper = mount(React.createElement(testCase.component, testCase.props));
      getWindow.mockReturnValue({
        document: {
          activeElement: jest.fn()
        }
      });
    });

    test('renders', () => {
      expect(
        shallow(React.createElement(testCase.component, testCase.props))
      ).toMatchSnapshot();
    });

    test('popover toggles on input focus/blur states', () => {
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );

      wrapper.find('input').last().props().onFocus();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        true
      );

      wrapper.find('input').last().props().onBlur();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );
    });

    test('popover toggles on button focus/blur states', () => {
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );

      wrapper.find('button').last().props().onFocus();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        true
      );

      wrapper.find('button').last().props().onBlur();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );
    });

    test('popover toggles on container mouseOver/mouseOut states', () => {
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );

      wrapper
        .find('[data-test="control-text-container"]')
        .props()
        .onMouseOver();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        true
      );

      wrapper.find('[data-test="control-text-container"]').props().onMouseOut();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );
    });
  });

  describe('Disabled onMouseOut behaviour when onFocus is set', () => {
    beforeEach(() => {
      testCase = testCases.errorInline;
      wrapper = mount(React.createElement(testCase.component, testCase.props));
      getWindow.mockReturnValue({
        document: {
          activeElement: jest.fn()
        }
      });
    });

    test('popover toggles on container mouseOver/mouseOut states', () => {
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        false
      );

      const inputEl = getWindow().document.activeElement;
      wrapper.instance().setInputElement(inputEl);
      wrapper.instance().onFocus();
      wrapper.update();

      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        true
      );

      wrapper.find('[data-test="control-text-container"]').props().onMouseOut();

      wrapper.update();
      expect(wrapper.find('[data-test="popover-component"]').exists()).toBe(
        true
      );
    });
  });

  describe(testCases.allOptions.description, () => {
    beforeEach(() => {
      testCase = testCases.allOptions;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('extra props works', () => {
      const props = wrapper.find('input').props();
      expect(props).toHaveProperty('autoFocus', true);
      expect(props).toHaveProperty('placeholder');
    });
  });
});
