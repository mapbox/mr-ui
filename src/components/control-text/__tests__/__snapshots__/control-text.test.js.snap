// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlText all options renders 1`] = `
<ControlWrapper
  id="testinput"
  themeControlWrapper="bg-gray-faint"
>
  <ControlLabel
    aside={
      <span>
        Aside text
      </span>
    }
    id="testinput"
    optional={true}
    text="All options"
    themeLabel="txt-s txt-bold"
  />
  <div
    className="flex-parent"
    data-test="control-text-container"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <input
      aria-invalid={true}
      aria-required={false}
      autoCapitalize="none"
      autoCorrect="off"
      autoFocus={true}
      className="txt-bold input round-l flex-child flex-child--grow"
      data-test="testinput-input"
      id="testinput"
      name="testinput"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Enter an email address"
      spellCheck={false}
      style={
        Object {
          "boxShadow": "inset 0 0 0 1px #f74e4e",
        }
      }
      type="email"
      value="foo@bar.baz"
    />
    <button
      aria-label="oh no!"
      className="flex-child bg-red color-white round-r px6"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      role="alert"
      type="button"
    >
      <span
        className="flex-parent flex-parent--center-cross flex-parent--center-main"
      >
        <Icon
          inline={false}
          name="alert"
          passthroughProps={Object {}}
          size={18}
          themeIcon="cursor-pointer"
        />
      </span>
    </button>
  </div>
</ControlWrapper>
`;

exports[`ControlText basic onChange gets called with input id and value 1`] = `
<ControlWrapper
  id="testinput"
  themeControlWrapper=""
>
  <ControlLabel
    id="testinput"
    optional={false}
    text="Basic"
    themeLabel="txt-s txt-bold mb6"
  />
  <input
    aria-required={true}
    className="input"
    data-test="testinput-input"
    id="testinput"
    name="testinput"
    onChange={[Function]}
    type="text"
    value=""
  />
</ControlWrapper>
`;

exports[`ControlText basic renders 1`] = `
<ControlWrapper
  id="testinput"
  themeControlWrapper=""
>
  <ControlLabel
    id="testinput"
    optional={false}
    text="Basic"
    themeLabel="txt-s txt-bold mb6"
  />
  <input
    aria-required={true}
    className="input"
    data-test="testinput-input"
    id="testinput"
    name="testinput"
    onChange={[Function]}
    type="text"
    value=""
  />
</ControlWrapper>
`;

exports[`ControlText default validation error renders 1`] = `
<ControlWrapper
  id="testinput"
  themeControlWrapper=""
  validationError="oh no!"
>
  <ControlLabel
    id="testinput"
    optional={false}
    text="Validation error"
    themeLabel="txt-s txt-bold mb6"
  />
  <input
    aria-invalid={true}
    aria-required={true}
    className="input"
    data-test="testinput-input"
    id="testinput"
    name="testinput"
    onChange={[Function]}
    type="text"
    value=""
  />
</ControlWrapper>
`;

exports[`ControlText disabled renders 1`] = `
<ControlWrapper
  id="testinput"
  themeControlWrapper=""
>
  <ControlLabel
    id="testinput"
    optional={false}
    text="Disabled"
    themeLabel="txt-s txt-bold mb6"
  />
  <input
    aria-required={true}
    className="input"
    data-test="testinput-input"
    disabled={true}
    id="testinput"
    name="testinput"
    onChange={[Function]}
    type="text"
    value=""
  />
</ControlWrapper>
`;

exports[`ControlText error inline option renders 1`] = `
<ControlWrapper
  id="testinput"
  themeControlWrapper=""
>
  <ControlLabel
    id="testinput"
    optional={false}
    text="Error inline"
    themeLabel="txt-s txt-bold mb6"
  />
  <div
    className="flex-parent"
    data-test="control-text-container"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <input
      aria-invalid={true}
      aria-required={true}
      className="input round-l flex-child flex-child--grow"
      data-test="testinput-input"
      id="testinput"
      name="testinput"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "boxShadow": "inset 0 0 0 1px #f74e4e",
        }
      }
      type="text"
      value=""
    />
    <button
      aria-label="oh no!"
      className="flex-child bg-red color-white round-r px6"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      role="alert"
      type="button"
    >
      <span
        className="flex-parent flex-parent--center-cross flex-parent--center-main"
      >
        <Icon
          inline={false}
          name="alert"
          passthroughProps={Object {}}
          size={18}
          themeIcon="cursor-pointer"
        />
      </span>
    </button>
  </div>
</ControlWrapper>
`;
